runtime/integration-tests/src/generic/cases/liquidity_pools.rs

```diff
@@ -0,0 +1,2371 @@
+	v3::{
+		AssetId, Fungibility, Junction, Junction::*, Junctions, Junctions::*, MultiAsset,
+		MultiAssets, MultiLocation, WeightLimit,
+	},
+	VersionedMultiAsset, VersionedMultiAssets, VersionedMultiLocation,
+};
+use xcm_executor::traits::Convert as C1;
+
+use crate::{
+	generic::{
+		config::Runtime,
+		env::{Blocks, Env},
+		envs::fudge_env::{handle::FudgeHandle, FudgeEnv, FudgeSupport},
+		utils::{genesis, genesis::Genesis},
+	},
+	utils::{accounts::Keyring, AUSD_CURRENCY_ID},
+};
+
+mod utils {
+	use super::*;
+
+	pub fn parachain_account(id: u32) -> AccountId {
+		polkadot_parachain::primitives::Sibling::from(id).into_account_truncating()
+	}
+
+	pub fn xcm_metadata(transferability: CrossChainTransferability) -> Option<XcmMetadata> {
+		match transferability {
+			CrossChainTransferability::Xcm(x) | CrossChainTransferability::All(x) => Some(x),
+			_ => None,
+		}
+	}
+
+	pub fn setup_xcm<T: Runtime + FudgeSupport>(env: &mut FudgeEnv<T>) {
+		env.parachain_state_mut(|| {
+			// Set the XCM version used when sending XCM messages to sibling.
+			assert_ok!(pallet_xcm::Pallet::<T>::force_xcm_version(
+				<T as frame_system::Config>::RuntimeOrigin::root(),
+				Box::new(MultiLocation::new(
+					1,
+					Junctions::X1(Junction::Parachain(T::FudgeHandle::SIBLING_ID)),
+				)),
+				XCM_VERSION,
+			));
+		});
+
+		env.sibling_state_mut(|| {
+			// Set the XCM version used when sending XCM messages to parachain.
+			assert_ok!(pallet_xcm::Pallet::<T>::force_xcm_version(
+				<T as frame_system::Config>::RuntimeOrigin::root(),
+				Box::new(MultiLocation::new(
+					1,
+					Junctions::X1(Junction::Parachain(T::FudgeHandle::PARA_ID)),
+				)),
+				XCM_VERSION,
+			));
+		});
+
+		env.relay_state_mut(|| {
+			assert_ok!(polkadot_runtime_parachains::hrmp::Pallet::<
+				FudgeRelayRuntime<T>,
+			>::force_open_hrmp_channel(
+				<FudgeRelayRuntime<T> as frame_system::Config>::RuntimeOrigin::root(),
+				Id::from(T::FudgeHandle::PARA_ID),
+				Id::from(T::FudgeHandle::SIBLING_ID),
+				10,
+				1024,
+			));
+
+			assert_ok!(polkadot_runtime_parachains::hrmp::Pallet::<
+				FudgeRelayRuntime<T>,
+			>::force_open_hrmp_channel(
+				<FudgeRelayRuntime<T> as frame_system::Config>::RuntimeOrigin::root(),
+				Id::from(T::FudgeHandle::SIBLING_ID),
+				Id::from(T::FudgeHandle::PARA_ID),
+				10,
+				1024,
+			));
+
+			assert_ok!(polkadot_runtime_parachains::hrmp::Pallet::<
+				FudgeRelayRuntime<T>,
+			>::force_process_hrmp_open(
+				<FudgeRelayRuntime<T> as frame_system::Config>::RuntimeOrigin::root(),
+				0,
+			));
+		});
+
+		env.pass(Blocks::ByNumber(1));
+	}
+}
+
+type FudgeRelayRuntime<T> = <<T as FudgeSupport>::FudgeHandle as FudgeHandle<T>>::RelayRuntime;
+
+use utils::*;
+
+mod altair {
+	use altair_runtime::{CurrencyIdConvert, PoolPalletIndex};
+
+	pub const KSM_ASSET_ID: CurrencyId = CurrencyId::ForeignAsset(1000);
+
+	use super::*;
+
+	mod utils {
+		use super::*;
+
+		pub fn register_ausd<T: Runtime + FudgeSupport>() {
```

Comment:
Just saw this. For these methods, you have a utility to define currencies in a way that later you can later refer to their properties easily in https://github.com/centrifuge/centrifuge-chain/blob/main/runtime/integration-tests/src/generic/utils/currency.rs

If you define AUSD there, for example, you can later write something like `Ausd::ED`

Reply:
Oh, that's pretty awesome, I guess I missed it while dealing with the test tornado ^^.